File: 04-estructuras-datos/14-integracion/java/ejemplos/GestorRecursos.java
import java.util.*;

public class GestorRecursos {
    record Recurso(String codigo, String titulo, String categoria) {}
    record Solicitud(String codigoRecurso, int prioridad) {}

    public static void main(String[] args) {
        List<Recurso> recursos = new ArrayList<>();
        Map<String, Recurso> porCodigo = new HashMap<>();
        Set<String> categorias = new HashSet<>();
        PriorityQueue<Solicitud> solicitudes = new PriorityQueue<>(Comparator.comparingInt(Solicitud::prioridad));

        Recurso recurso = new Recurso("R-001", "Guía de Java", "Programación");
        recursos.add(recurso);                    // List conserva el catálogo principal.
        porCodigo.put(recurso.codigo(), recurso); // Map funciona como índice por código.
        categorias.add(recurso.categoria());      // Set evita categorías repetidas.
        solicitudes.offer(new Solicitud("R-001", 1)); // PriorityQueue ordena las solicitudes por prioridad.

        System.out.println("Recurso: " + porCodigo.get("R-001"));
        System.out.println("Solicitud: " + solicitudes.poll());
    }
}

File: 04-estructuras-datos/14-integracion/scala/ejemplos/GestorRecursos.scala
import scala.collection.mutable.PriorityQueue

case class Recurso(codigo: String, titulo: String, categoria: String)
case class Solicitud(codigoRecurso: String, prioridad: Int)

object GestorRecursos:
  def main(args: Array[String]): Unit =
    val recursos = Vector(Recurso("R-001", "Guía de Scala", "Programación"))
    val porCodigo = recursos.map(r => r.codigo -> r).toMap
    val categorias = recursos.map(_.categoria).toSet
    given Ordering[Solicitud] = Ordering.by(s => -s.prioridad)
    val solicitudes = PriorityQueue(Solicitud("R-001", 1))
    // Map indexa por código y PriorityQueue organiza las solicitudes por prioridad.
    println(porCodigo("R-001"))
    println(categorias)
    println(solicitudes.dequeue())
